Allow remote playback device to choose the remote playback source.

markafoltz opened the pull request:
This addresses Issue #125: RemotePlaybackState enum can become misleading when changing media.src.

It clarifies that:
1. All media resources in a media element are eligible for remote playback
1. Metadata about all the media resources may be sent to the remote playback device, so it can choose the remote playback source

Some more clarification could be done: 

* Adding a step to 5.2.4 ("Establishing a connection with a remote playback device") to indicate when the selection of the remote playback source happens.  (Instead, I updated a non-normative note at the end of these steps.)
* Clarifying what happens when remote playback is disconnected, i.e. does the user agent automatically revert to the `.currentSrc` that was playing before remote playback was initiated.

Let me know if you want further changes along these lines.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/remote-playback/pull/141.html" title="Last updated on Nov 5, 2020, 9:44 PM UTC (852e47e)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/remote-playback/141/b0ecc12...852e47e.html" title="Last updated on Nov 5, 2020, 9:44 PM UTC (852e47e)">Diff</a>

markafoltz committed:
Allow remote playback device to choose the remote playback source.

markafoltz commented:
PTAL @mounirlamouri @jernoble

baylesj commented on index.html:
```diff
@@ -571,24 +571,27 @@ <h5>
+                playback source</a>.
+            </p>
+            <p class="note">
+              The user agent may send metadata about all resources in
```
Is this metadata defined anywhere and linkable?

markafoltz replied:
OSP sends the [extended mime type](https://html.spec.whatwg.org/multipage/embedded-content.html#attr-source-type) for each resource, but it's up to the implementation what information to exchange with the remote playback device.  Can edit a bit to clarify what is meant here.

markafoltz replied:
Added link to mime type as an example.

baylesj commented on index.html:
```diff
@@ -1131,14 +1134,14 @@ <h4>
-            playback device and receiving media playback state in order to keep
-            the <a>media element state</a> and <a>remote playback state</a> in
-            sync (unless <a>media mirroring</a> is used).
+            The mechanisms that are used to connect the user agent with the
```
Nit: first sentence is confusing. Maybe add an oxford comma?

> The mechanisms that are used to connect the user agent with the
<a>remote playback device</a>, select the <a>remote playback source</a>**,**
and initiate playback are **all** **implemented by** the user agent.

markafoltz replied:
Done.

baylesj commented on index.html:
```diff
@@ -1131,14 +1134,14 @@ <h4>
+            The mechanisms that are used to connect the user agent with the
+            <a>remote playback device</a>, select the <a>remote playback
+            source</a> and initiate playback are implementation choices of the
+            user agent. The connection will likely have to provide a two-way
```
Nit: The connection provides a two way messaging abstraction? Isn't that the connection itself, and the user agent and remote playback device have to provide the messaging abstractoin?

markafoltz replied:
I think the "connection"  refers in an abstract way with the communication mechanism, which can be used for syncing media element state or other purposes.  As this was pre-existing text in the spec, I'm inclined to leave it alone in this PR.

baylesj reviewed (commented):
Not sure if the behavior on remote playback end is implementation specific or not, so if we expect that to be standardized (versus the implementation-specific playback source selection mechanism) I'd like to see that in this document.

I think the note for 5.2.4 is fine as is.

markafoltz committed:
Respond to comments from baylesj

mounirlamouri commented on index.html:
```diff
@@ -571,24 +571,28 @@ <h5>
+              the <a>availability sources set</a> as a potential <a>remote
+                playback source</a>.
+            </p>
+            <p class="note">
```
Could we imagine having this note as a RECOMMENDED instead? Something like:

```
If the user agent cannot determine a <a>remote playback source</a> appropriate for the <a>remote playback device</a>, it is RECOMMENDED that the user agent send metadata [...] to the <a>remote playback device</a> so it can run its own <a>resource selection algorithm</a> and choose the <a>remote playback source</a>.
```

jernoble commented:
LGTM.

anssiko commented:
It seems @jernoble LGTM'd this PR conditional to @mounirlamouri's proposal https://github.com/w3c/remote-playback/pull/141#discussion_r508217840 baked in?

@mfoltzgoogle feel free to update the PR per the proposal and try to get quick confirmation from @mounirlamouri and/or @baylesj and @jernoble  prior to merge.

markafoltz committed:
Update per mlamouri comment.  Move note.

markafoltz commented:
Updated the PR per the proposal from @mounirlamouri.

I also moved a separate note regarding source selection to this part of the spec, so that all the discussion of source selection is now in this section.

@mounirlamouri @baylesj @jernoble PTAL

markafoltz committed:
Rewrap
